aiohttp/client_ws.py

```diff
@@ -85,48 +86,62 @@ def __init__(
         self._reset_heartbeat()
 
     def _cancel_heartbeat(self) -> None:
-        if self._pong_response_cb is not None:
-            self._pong_response_cb.cancel()
-            self._pong_response_cb = None
-
+        self._cancel_pong_response_cb()
         if self._heartbeat_cb is not None:
             self._heartbeat_cb.cancel()
             self._heartbeat_cb = None
 
+    def _cancel_pong_response_cb(self) -> None:
+        if self._pong_response_cb is not None:
+            self._pong_response_cb.cancel()
+            self._pong_response_cb = None
+
     def _reset_heartbeat(self) -> None:
-        self._cancel_heartbeat()
-
-        if self._heartbeat is not None:
-            self._heartbeat_cb = call_later(
-                self._send_heartbeat,
-                self._heartbeat,
-                self._loop,
-                timeout_ceil_threshold=(
-                    self._conn._connector._timeout_ceil_threshold
-                    if self._conn is not None
-                    else 5
-                ),
-            )
+        if self._heartbeat is None:
+            return
+        self._cancel_pong_response_cb()
+        assert self._loop is not None
+        loop = self._loop
+        conn = self._conn
+        timeout_ceil_threshold = (
+            conn._connector._timeout_ceil_threshold if conn is not None else 5
+        )
+        now = loop.time()
+        when = calculate_timeout_when(now, self._heartbeat, timeout_ceil_threshold)
+        self._heartbeat_when = when
+        if self._heartbeat_cb is None:
+            # We do not cancel the previous heartbeat_cb here because
+            # it generates a significant amount of TimerHandle churn
+            # which causes asyncio to rebuild the heap frequently.
+            # Instead _send_heartbeat() will reschedule the next
+            # heartbeat if it fires too early.
+            self._heartbeat_cb = loop.call_at(when, self._send_heartbeat)
 
     def _send_heartbeat(self) -> None:
-        if self._heartbeat is not None and not self._closed:
-            # fire-and-forget a task is not perfect but maybe ok for
-            # sending ping. Otherwise we need a long-living heartbeat
-            # task in the class.
-            self._loop.create_task(self._writer.ping())  # type: ignore[unused-awaitable]
-
-            if self._pong_response_cb is not None:
-                self._pong_response_cb.cancel()
-            self._pong_response_cb = call_later(
-                self._pong_not_received,
-                self._pong_heartbeat,
-                self._loop,
-                timeout_ceil_threshold=(
-                    self._conn._connector._timeout_ceil_threshold
-                    if self._conn is not None
-                    else 5
-                ),
+        self._heartbeat_cb = None
+        if self._heartbeat is None or self._closed:
+            return
+        loop = self._loop
+        now = loop.time()
+        if now < self._heartbeat_when:
```

Comment:
So, basically instead of cancelling we just check if we're running too early and then schedule a new one?
This feels like a bit of an anti-pattern and makes me wonder if asyncio can't be improved to make cancelling lightweight. It seems very counter-intuitive to think that it's faster to run a callback than to not run it...

Reply:
Wouldn't using heappush() rather than append() then heapify() be faster? Though probably still not significant enough...

Reply:
> This feels like a bit of an anti-pattern and makes me wonder if asyncio can't be improved to make cancelling lightweight. It seems very counter-intuitive to think that it's faster to run a callback than to not run it...

asyncio is optimized to assume timer handlers will fire which I think makes sense. Timer handles don't actually get get removed from the heap when they are cancelled and instead get marked as cancelled, and when they come up they are skipped as its too expensive otherwise. When they are being canceled/scheduled frequently its almost always better to let them fire and reschedule since there is no scheduling churn for the duration of the timer.

https://github.com/python/cpython/blob/1422500d020bd199b26357fc387f8b79b82226cd/Lib/asyncio/events.py#L158 https://github.com/python/cpython/blob/1422500d020bd199b26357fc387f8b79b82226cd/Lib/asyncio/events.py#L74

Reply:
Yep, thinking about it more, this is part of the low-level API that most users shouldn't interact with. So, I guess we should be dealing with this limitation.

Other places that might be improved:
client_proto.py:read_timeout: We cancel and reschedule everytime data is received.
client_ws.py: Do we need to deal with _pong_response_cb too? Cancelled and rescheduled on each heartbeat send.
web_protocol.py:.keep_alive(): Should we cancel if the val is already the same and a handle already exists?

Maybe some issues in connector.py if there are many connections that are getting closed.

Also, helpers.py:TimeoutHandle is only used once in client.py and seems like something that could be replaced by asyncio.timeout(). We'd need to use asyncio.Timeout.reschedule() in the resp.connection callback.

Reply:
> Also, helpers.py:TimeoutHandle is only used once in client.py and seems like something that could be replaced by asyncio.timeout(). We'd need to use asyncio.Timeout.reschedule() in the resp.connection callback.

.reschedule() still cancels a TimerHandle though. As that's part of the high-level API of asyncio, I wonder if it should be changed in a similar way to reduce cancellations.

Reply:
> Yep, thinking about it more, this is part of the low-level API that most users shouldn't interact with. So, I guess we should be dealing with this limitation.
> 
> Other places that might be improved: 

> client_proto.py:read_timeout: We cancel and reschedule everytime data is received. 

I think it might be worth doing. It fires far less often than websockets though. Need to do some more profiling with this one.

> client_ws.py: Do we need to deal with _pong_response_cb too? Cancelled and rescheduled on each heartbeat send. 

This is a rare case where its almost never rescheduled since we expect it to respond right away or be closed.  Its probably not worth optimizing.

> web_protocol.py:.keep_alive(): Should we cancel if the val is already the same and a handle already exists?
> 

This one doesn't seem to get rescheduled frequently when I looked at the profile.

> Maybe some issues in connector.py if there are many connections that are getting closed.
> 

cleanup_closed could use some improvements.  Something to investigate for sure.

> Also, helpers.py:TimeoutHandle is only used once in client.py and seems like something that could be replaced by asyncio.timeout(). We'd need to use asyncio.Timeout.reschedule() in the resp.connection callback.

That looks like a good opportunity to cleanup there.

Reply:
https://github.com/aio-libs/aiohttp/issues/8613

Reply:
The cleanup closed one is likely work prioritizing as it tends to create a lot of timer handlers, especially with the 2s timeout.